cvmfs/swissknife_filestats.cc

```diff
@@ -0,0 +1,317 @@
+  num_downloaded_ = 0;
+
+  bool success = false;
+  if (IsHttpUrl(repo_url)) {
+    const bool follow_redirects = false;
+    if (!this->InitDownloadManager(follow_redirects) ||
+        !this->InitVerifyingSignatureManager(repo_keys)) {
+      LogCvmfs(kLogCatalog, kLogStderr, "Failed to init remote connection");
+      return 1;
+    }
+
+    HttpObjectFetcher<catalog::Catalog,
+                      history::SqliteHistory> fetcher(repo_name,
+                                                      repo_url,
+                                                      tmp_dir,
+                                                      download_manager(),
+                                                      signature_manager());
+    success = Run(&fetcher);
+  } else {
+    LocalObjectFetcher<> fetcher(repo_url, tmp_dir);
+    success = Run(&fetcher);
+  }
+
+  return (success) ? 0 : 1;
+}
+
+template <class ObjectFetcherT>
+bool CommandFileStats::Run(ObjectFetcherT *object_fetcher)
+{
+  running_ = true;
+
+  string abs_path = GetAbsolutePath(db_path_);
+  unlink(abs_path.c_str());
+  db_ = FileStatsDatabase::Create(db_path_);
+  db_->InitStatements();
+
+  RemoveTree(tmp_db_path_);
+  assert(MkdirDeep(tmp_db_path_, 0755));
+
+  typename CatalogTraversal<ObjectFetcherT>::Parameters params;
+  params.object_fetcher = object_fetcher;
+  CatalogTraversal<ObjectFetcherT> traversal(params);
+  traversal.RegisterListener(&CommandFileStats::CatalogCallback, this);
+
+  pthread_create(&thread_processing_, NULL, MainProcessing, this);
+
+  bool ret = traversal.Traverse();
+
+  running_ = false;
+  pthread_join(thread_processing_, NULL);
+
+  db_->DestroyStatements();
+
+  return ret;
+}
+
+void CommandFileStats::CatalogCallback(
+  const CatalogTraversalData<catalog::Catalog> &data) {
+  string out_path =  tmp_db_path_ + StringifyInt(num_downloaded_ + 1) + ".db";
+  assert(CopyPath2Path(data.catalog->database_path(), out_path));
+  ++num_downloaded_;
+}
+
+void *CommandFileStats::MainProcessing(void *data) {
+  CommandFileStats *repo_stats = static_cast<CommandFileStats *>(data);
+  int processed = 0;
+  while (repo_stats->running_ || processed < repo_stats->num_downloaded_) {
+    if (processed < repo_stats->num_downloaded_) {
+      LogCvmfs(kLogCatalog, kLogStdout, "Processing catalog %d", processed);
+      string db_path = repo_stats->tmp_db_path_ + "/" +
+                       StringifyInt(processed + 1) + ".db";
+      repo_stats->ProcessCatalog(db_path);
+      unlink(db_path.c_str());
+      ++processed;
+    }
+  }
+  return NULL;
+}
+
+
+
+void CommandFileStats::ProcessCatalog(string db_path) {
+  sqlite::Database<catalog::CatalogDatabase> *cat_db;
+  cat_db = sqlite::Database<catalog::CatalogDatabase>::Open(
+           db_path,
+           sqlite::Database<catalog::CatalogDatabase>::kOpenReadOnly);
+  int64_t file_size = GetFileSize(db_path);
+  sqlite::Sql *catalog_count = new sqlite::Sql(cat_db->sqlite_db(),
```

Comment:
This can be stack allocated

Reply:
The sql statement needs to be destroyed early (before closing the database). I think this is the easiest way.